04-data-structures-part1.Rmd

```diff
@@ -2,336 +2,271 @@
-hood, they are really just atomic vectors, with added dimension attributes.
+## Factors
 
-We can create one with the `matrix` function. Let's generate some random data:
+We said that columns in data.frames were vectors:
 
 ```{r}
-set.seed(1) # make sure the random numbers are the same for each run
-x <- matrix(rnorm(18), ncol=6, nrow=3)
-x
+str(cats$weight)
+str(cats$likes_string)
 ```
 
+But what about
+
 ```{r}
-str(x)
+str(cats$coat)
 ```
 
-You can use `rownames`, `colnames`, and `dimnames` to set or
-retrieve the column and rownames of a matrix. The functions `nrow` and `ncol`
-will tell you the number of rows and columns (this also applies to data frames!),
-while `length` will tell you the number of elements.
-
->
-> ## Challenge 3 {.challenge}
->
-> What do you think will be the result of
-> `length(x)`?
-> Try it.
-> Were you right? Why / why not?
->
-
->
-> ## Challenge 4 {.challenge}
->
-> Make another matrix, this time containing the numbers 1:50,
-> with 5 columns and 10 rows.
-> Did the `matrix` function fill your matrix by column, or by
-> row, as its default behaviour?
-> See if you can figure out how to change this.
-> (hint: read the documentation for `matrix`!)
->
-
-## Factors
+Another important data structure is called a *factor*. Factors usually look like character data, but are typically used to represent categorical information. For example, let's make a vector of strings labeling cat colorations for all the cats in our study:
 
-Factors are special vectors that represent categorical data. Factors can be
-ordered or unordered and are important when for modeling functions such as
-`aov()`, `lm()` and `glm()` and also in plot methods.
+```{r}
+coats <- c('tabby', 'tortoiseshell', 'tortoiseshell', 'black', 'tabby')
+coats
+str(coats)
+```
 
-Factors can only contain predefined values, and we can create one with the
-`factor` function:
+We can turn a vector into a factor like so:
 
 ```{r}
-x <- factor(c("yes", "no", "no", "yes", "yes"))
-x
+CATegories <- factor(coats)
+str(CATegories)
 ```
 
-So we can see that the output is very similar to a character vector, but with an
-attached levels component. This becomes clearer when we look at its structure:
+Now R has noticed that there are three possible categories in our data - but it also did something surprising; instead of printing out the strings we gave it, we got a bunch of numbers instead. R has replaced our human-readable categories with numbered indices under the hood:
 
 ```{r}
-str(x)
+typeof(coats[1])
+typeof(CATegories[1])
 ```
 
-This reveals something important: while factors look (and often behave) like
-character vectors, they are actually integers under the hood, and here, we can
-see that "no" is represented by a 1, and "yes" a 2.
+> ## Challenge 2 {.challenge}
+> When we loaded our `cats` data, the coats column was interpreted as a factor; 
+> try using the help for `read.table` to figure out how to keep text columns as character vectors instead of factors; 
```

Comment:
should this be `?read.table` or is part of the exercise to help them remember how to do that.